File: src/cedict_parser.py
# Parser that extracts lines from the CEDICT Chinese-English dictionary and
# exports them as a JSON file.

import json
from parse import parse

# Assumes the parser is run from the src directory
INPUT_CEDICT_PATH = '../public/cedict_ts.u8'

OUTPUT_MAPPINGS_PATH = '../public/char-mappings.json'
OUTPUT_DICTIONARY_PATH = '../public/cedict-ts.json'


def parse_line(line: str) -> dict:
    """Converts a line from the CEDICT into a dictionary along its delimiters.
        The basic format of each line is specified by CEDICT documentation as:
        "traditional simplified [pin1 yin1] /gloss; gloss; .../gloss; gloss; .../"

    Args:
        line (str): A line from the CEDICT file.

    Returns:
        dict: A dictionary with keys "traditional," "simplified", "pinyin", and
            "glossary".
    """

    parsed_line = list(parse('{} {} [{}] /{}/', line))
    parsed_line[-1] = parsed_line[-1].split('/')

    keys = ['traditional', 'simplified', 'pinyin', 'glossary']
    phrase = {key: value for (key, value) in zip(keys, parsed_line)}
    phrase['pinyin'] = phrase['pinyin'].lower()

    return phrase


if __name__ == '__main__':
    # Index mappings for simplified and traditional characters onto the same
    # dictionary entry
    char_mapping = {}
    dict_entries = []

    # Open CEDICT file
    with open(INPUT_CEDICT_PATH, encoding='utf8') as cedict:
        for line in cedict:
            # Inserts phrase into dictionary and creates index mappings for characters
            phrase = parse_line(line.strip())

            # Mappings store a set of indices of dictionary entries for terms
            # with multiple definitions
            if phrase['traditional'] not in char_mapping:
                char_mapping[phrase['traditional']] = set()

            if phrase['simplified'] not in char_mapping:
                char_mapping[phrase['simplified']] = set()

            char_mapping[phrase['traditional']].add(len(dict_entries))
            char_mapping[phrase['simplified']].add(len(dict_entries))

            # Adds phrase to the dictionary
            dict_entries.append(phrase)

    # Convert sets to lists for JSON serialization
    for key in char_mapping:
        char_mapping[key] = list(char_mapping[key])

    # Outputs character mappings and dictionary as json files
    with (open(OUTPUT_MAPPINGS_PATH, 'w', encoding='utf-8') as mappings_json,
            open(OUTPUT_DICTIONARY_PATH, 'w', encoding='utf-8') as dict_json):
        json.dump(char_mapping, mappings_json, ensure_ascii=False)
        json.dump(dict_entries, dict_json, ensure_ascii=False)

File: src/cedict_parser.js
import fs from 'fs';
import readline from 'readline';
import { convert } from 'pinyin-pro';

// Assumes the parser is run from the root directory
const INPUT_CEDICT_PATH = 'public/cedict_ts.u8';
const OUTPUT_MAPPINGS_PATH = 'public/char-mappings.json';
const OUTPUT_DICTIONARY_PATH = 'public/cedict-ts.json';

/**
 * Converts a line from the CEDICT into a dictionary along its delimiters.
 * The basic format of each line is specified by CEDICT documentation as:
 * "traditional simplified [pin1 yin1] /gloss; gloss; .../gloss; gloss; .../"
 *
 * @param {string} line A line from the CEDICT file
 * @returns {object} An JSON object containing each word with their corresponding
 * definitions
 */
function parseLine(line) {
  const match = line.match(/^(\S+) (\S+) \[([^\]]+)\] \/(.+)\/$/);
  if (!match) {
    throw Error(`Encountered unexpected line in CEDICT file: \n\t${line}`);
  }

  const [_, traditional, simplified, rawPinyin, rawGlossary] = match;
  const glossary = rawGlossary.split('/');

  // CEDICT uses "u:" instead of "v" for ü
  let pinyin = rawPinyin.replace(/u:/g, 'v').toLowerCase();

  // Converts tone numbers (e.g., "pin1 yin1") to diacritics (e.g., "pīn yīn")
  pinyin = convert(pinyin);

  return {
    traditional,
    simplified,
    pinyin,
    glossary,
  };
}

(async () => {
  const charMapping = {};
  const dictEntries = [];

  // Create a read stream to handle CEDICT file
  const fileStream = fs.createReadStream(INPUT_CEDICT_PATH, {
    encoding: 'utf8',
  });

  const readStream = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  });

  for await (const line of readStream) {
    // Skip empty lines and comment lines
    const trimmedLine = line.trim();
    if (!trimmedLine || trimmedLine.startsWith('#')) {
      continue;
    }

    // Each traditional and simpified character will map to `entryIdx` of the
    // dictionary
    const phrase = parseLine(trimmedLine);
    const entryIdx = dictEntries.length;

    // Mappings store a set of indices of dictionary entries for terms with
    // multiple definitions
    if (!(phrase.traditional in charMapping)) {
      charMapping[phrase.traditional] = new Set();
    }

    if (!(phrase.simplified in charMapping)) {
      charMapping[phrase.simplified] = new Set();
    }

    charMapping[phrase.traditional].add(entryIdx);
    charMapping[phrase.simplified].add(entryIdx);

    dictEntries.push(phrase);
  }

  for (const key in charMapping) {
    charMapping[key] = Array.from(charMapping[key]);
  }

  // Output character mappings and dictionary as JSON files
  try {
    fs.writeFileSync(
      OUTPUT_MAPPINGS_PATH,
      JSON.stringify(charMapping, null),
      'utf8'
    );

    fs.writeFileSync(
      OUTPUT_DICTIONARY_PATH,
      JSON.stringify(dictEntries, null),
      'utf8'
    );

    console.log('Parsing complete.');
  } catch (err) {
    console.error('Error writing JSON output files: ', err);
  }
})()
